neurom/core/soma.py

```diff
@@ -267,20 +202,238 @@ def __str__(self):
+        SomaType.SOMA_SIMPLE_CONTOUR: _soma_simple_contour_radius,
+    }[morphio_soma.type]
+
+    return soma_algo(morphio_soma)
+
+def soma_area(soma):
+
+    if hasattr(soma, "to_morphio"):
+        morphio_soma = soma.to_morphio()
+    else:
+        morphio_soma = soma
+
+    def _soma_single_point_area(morphio_soma):
+        return 4. * math.pi * soma_radius(morphio_soma)**2
+
+    def _soma_undefined_area(morphio_soma):
+        warnings.warn('Approximating soma area by a sphere. {}'.format(soma))
+        return _soma_single_point_radius(morphio_soma)
+
+    def _soma_cylinders_area(morphio_soma):
+        points = np.concatenate(
+            (
+                morphio_soma.points,
+                0.5 * morphio_soma.diameters[:, np.newaxis],
+            ),
+            axis=1,
+        )
+        return sum(
+            morphmath.segment_area((p0, p1)) for p0, p1 in zip(points, points[1:])
+        )
+
+    def _soma_three_point_cylinders_area(morphio_soma):
+        r = 0.5 * morphio_soma.diameters[0]
+        h = morphmath.point_dist(morphio_soma.points[1], morphio_soma.points[2])
+        return 2.0 * math.pi * r * h  # ignores the 'end-caps' of the cylinder
+
+    soma_algo = {
+        SomaType.SOMA_UNDEFINED: _soma_undefined_area,
+        SomaType.SOMA_SINGLE_POINT: _soma_single_point_area,
+        SomaType.SOMA_CYLINDERS: _soma_cylinders_area,
+        SomaType.SOMA_NEUROMORPHO_THREE_POINT_CYLINDERS: _soma_three_point_cylinders_area,
+        SomaType.SOMA_SIMPLE_CONTOUR: _soma_single_point_area,
+    }[morphio_soma.type]
+
+    return soma_algo(morphio_soma)
+
+
+def soma_volume(soma):
+    """Calculate soma volume."""
+
+    if hasattr(soma, "to_morphio"):
+        morphio_soma = soma.to_morphio()
+    else:
+        morphio_soma = soma
+
+    def _soma_single_point_volume(morphio_soma):
+        return 4.0 / 3 * math.pi * soma_radius(morphio_soma)**3
+
+    def _soma_undefined_volume(morphio_soma):
+        warnings.warn('Approximating soma volume by a sphere. {}'.format(soma))
+        return _soma_single_point_volume(morphio_soma)
+
+    def _soma_cylinders_volume(soma):
+        points = np.concatenate(
+            (
+                morphio_soma.points,
+                0.5 * morphio_soma.diameters[:, np.newaxis],
+            ),
+            axis=1,
+        )
+        return sum(
+            morphmath.segment_volume((p0, p1)) for p0, p1 in zip(points, points[1:])
+        )
+
+    def _soma_three_point_cylinders_volume(morphio_soma):
+        return 2. * math.pi * soma_radius(morphio_soma)**3
+
+
+    soma_algo = {
+        SomaType.SOMA_UNDEFINED: _soma_undefined_volume,
+        SomaType.SOMA_SINGLE_POINT: _soma_single_point_volume,
+        SomaType.SOMA_CYLINDERS: _soma_cylinders_volume,
+        SomaType.SOMA_NEUROMORPHO_THREE_POINT_CYLINDERS: _soma_three_point_cylinders_volume,
+        SomaType.SOMA_SIMPLE_CONTOUR: _soma_undefined_volume,
+    }[morphio_soma.type]
```

Comment:
Isn't it better to define these functions and dicts outside the function?

Reply:
Moved the functions outside. I however left the dicts inside.